File: apps/api/internal/modules/articles/paywall.go
// Package articles — module articles + troncature paywall (zero-leak).
package articles

import "strings"

// Visibilités (miroir de l'enum ContentVisibility).
const (
	VisPublic          = "PUBLIC"
	VisMembersOnly     = "MEMBERS_ONLY"
	VisPaidSubscribers = "PAID_SUBSCRIBERS"
	VisTierSpecific    = "TIER_SPECIFIC"
)

// UserEntitlements décrit les droits du lecteur (miroir TS).
type UserEntitlements struct {
	IsMember         bool
	IsPaidSubscriber bool
	TierID           *string
}

// PaywallMeta décrit la coupure appliquée.
type PaywallMeta struct {
	Visibility            string  `json:"visibility"`
	TeaserParagraphsCount int     `json:"teaserParagraphsCount"`
	RequiredTierID        *string `json:"requiredTierId"`
	TotalLengthBytes      int     `json:"totalLengthBytes"`
	PreviewLengthBytes    int     `json:"previewLengthBytes"`
}

// PaywallCutResult est le résultat de la troncature.
type PaywallCutResult struct {
	Content       string       `json:"content"`
	IsTruncated   bool         `json:"isTruncated"`
	AccessGranted bool         `json:"accessGranted"`
	PaywallMeta   *PaywallMeta `json:"paywallMeta"`
}

// Marqueurs de coupure des différents éditeurs (Ghost, Lexical, TipTap…).
var paywallMarkers = []string{
	"<!--members-only-->",
	"<!--paywall-->",
	"<!--kg-gated-block:begin-->",
	"<!--qoe-paywall-->",
	`data-node-type="paywall"`,
	`data-node-type="paywall-divider"`,
	`data-type="paywall-divider"`,
	`class="qoe-paywall-divider"`,
	`class="paywall-divider"`,
}

// CheckContentAccess vérifie si les droits du lecteur suffisent.
func CheckContentAccess(visibility string, e UserEntitlements, requiredTierID *string) bool {
	switch visibility {
	case VisPublic, "":
		return true
	case VisMembersOnly:
		return e.IsMember || e.IsPaidSubscriber
	case VisPaidSubscribers:
		return e.IsPaidSubscriber
	case VisTierSpecific:
		if !e.IsPaidSubscriber {
			return false
		}
		if requiredTierID == nil {
			return true
		}
		return e.TierID != nil && *e.TierID == *requiredTierID
	}
	return true
}

// SliceContentAtPaywall applique la troncature (miroir TS) : jamais de contenu
// payant au-delà du marqueur transmis au lecteur non autorisé.
func SliceContentAtPaywall(rawContent string, e UserEntitlements, visibility string, requiredTierID *string) PaywallCutResult {
	totalBytes := len(rawContent)
	access := CheckContentAccess(visibility, e, requiredTierID)

	if access || rawContent == "" {
		return PaywallCutResult{Content: rawContent, AccessGranted: true}
	}

	// Cherche le premier marqueur.
	paywallIndex := -1
	for _, marker := range paywallMarkers {
		if idx := strings.Index(rawContent, marker); idx != -1 && (paywallIndex == -1 || idx < paywallIndex) {
			paywallIndex = idx
		}
	}

	var preview string
	teaserCount := 0

	if paywallIndex != -1 {
		preview = strings.TrimSpace(rawContent[:paywallIndex])
		teaserCount = strings.Count(preview, "</p>")
		if teaserCount == 0 {
			teaserCount = 1
		}
	} else {
		// Fallback : 2 premiers paragraphes, sinon 1, sinon 500 chars.
		closings := paragraphIndexes(rawContent)
		if len(closings) >= 2 {
			preview = strings.TrimSpace(rawContent[:closings[1]+4])
			teaserCount = 2
		} else if len(closings) == 1 {
			preview = strings.TrimSpace(rawContent[:closings[0]+4])
			teaserCount = 1
		} else {
			cut := len(rawContent)
			if cut > 500 {
				cut = 500
			}
			preview = strings.TrimSpace(rawContent[:cut])
			teaserCount = 1
		}
	}

	return PaywallCutResult{
		Content:       preview,
		IsTruncated:   true,
		AccessGranted: false,
		PaywallMeta: &PaywallMeta{
			Visibility:            visibility,
			TeaserParagraphsCount: teaserCount,
			RequiredTierID:        requiredTierID,
			TotalLengthBytes:      totalBytes,
			PreviewLengthBytes:    len(preview),
		},
	}
}

// paragraphIndexes retourne les positions des fermetures </p>.
func paragraphIndexes(s string) []int {
	var out []int
	offset := 0
	for {
		idx := strings.Index(s[offset:], "</p>")
		if idx == -1 {
			return out
		}
		abs := offset + idx
		out = append(out, abs)
		offset = abs + 4
	}
}

File: packages/utils/src/paywall.ts
import { ContentVisibility } from '@qoe/config';

export interface UserEntitlements {
  isMember: boolean; // Registered free subscriber (has provided email)
  isPaidSubscriber: boolean; // Active paid subscriber
  tierId?: string | null; // Optional specific subscription tier ID
}

export interface PaywallMeta {
  visibility: ContentVisibility;
  teaserParagraphsCount: number;
  requiredTierId?: string | null;
  totalLengthBytes: number;
  previewLengthBytes: number;
}

export interface PaywallCutResult {
  content: string;
  isTruncated: boolean;
  accessGranted: boolean;
  paywallMeta: PaywallMeta | null;
}

// Markers for paywall breaks across various editor outputs (Ghost, Lexical, TipTap, HTML comments)
const PAYWALL_MARKERS = [
  '<!--members-only-->',
  '<!--paywall-->',
  '<!--kg-gated-block:begin-->',
  '<!--qoe-paywall-->',
  'data-node-type="paywall"',
  'data-node-type="paywall-divider"',
  'data-type="paywall-divider"',
  'class="qoe-paywall-divider"',
  'class="paywall-divider"',
];

/**
 * Checks whether user entitlements meet the required content visibility rules.
 */
export function checkContentAccess(
  visibility: ContentVisibility = ContentVisibility.PUBLIC,
  entitlements: UserEntitlements,
  requiredTierId?: string | null
): boolean {
  if (visibility === ContentVisibility.PUBLIC) {
    return true;
  }

  if (visibility === ContentVisibility.MEMBERS_ONLY) {
    return entitlements.isMember || entitlements.isPaidSubscriber;
  }

  if (visibility === ContentVisibility.PAID_SUBSCRIBERS) {
    return entitlements.isPaidSubscriber;
  }

  if (visibility === ContentVisibility.TIER_SPECIFIC) {
    if (!entitlements.isPaidSubscriber) return false;
    if (!requiredTierId) return true;
    return entitlements.tierId === requiredTierId;
  }

  return true;
}

/**
 * Server-side zero-leak AST/HTML paywall truncation engine.
 * Ensures paid content bytes above the paywall marker are NEVER transmitted over the wire
 * to non-authorized visitors.
 */
export function sliceContentAtPaywall(
  rawContent: string,
  entitlements: UserEntitlements,
  visibility: ContentVisibility = ContentVisibility.PUBLIC,
  requiredTierId?: string | null
): PaywallCutResult {
  const totalLengthBytes = Buffer.byteLength(rawContent || '', 'utf-8');
  const accessGranted = checkContentAccess(visibility, entitlements, requiredTierId);

  // If access is granted, return full content
  if (accessGranted || !rawContent) {
    return {
      content: rawContent || '',
      isTruncated: false,
      accessGranted: true,
      paywallMeta: null,
    };
  }

  // Search for the first matching paywall marker
  let paywallIndex = -1;

  for (const marker of PAYWALL_MARKERS) {
    const index = rawContent.indexOf(marker);
    if (index !== -1 && (paywallIndex === -1 || index < paywallIndex)) {
      paywallIndex = index;
    }
  }

  let previewContent = '';
  let teaserParagraphsCount = 0;

  if (paywallIndex !== -1) {
    // Slice cleanly at the marker
    previewContent = rawContent.slice(0, paywallIndex).trim();
    // Count paragraph blocks in the preview
    const paragraphMatches = previewContent.match(/<\/p>/g);
    teaserParagraphsCount = paragraphMatches ? paragraphMatches.length : 1;
  } else {
    // Fallback: If no paywall marker is found in gated content, truncate to first 2 paragraphs
    const paragraphMatches = Array.from(rawContent.matchAll(/<\/p>/g));
    if (paragraphMatches.length >= 2 && paragraphMatches[1].index) {
      const cutEnd = paragraphMatches[1].index + 4; // Include </p>
      previewContent = rawContent.slice(0, cutEnd).trim();
      teaserParagraphsCount = 2;
    } else if (paragraphMatches.length === 1 && paragraphMatches[0].index) {
      const cutEnd = paragraphMatches[0].index + 4;
      previewContent = rawContent.slice(0, cutEnd).trim();
      teaserParagraphsCount = 1;
    } else {
      // If no paragraph tags, take first 500 characters
      previewContent = rawContent.slice(0, 500).trim();
      teaserParagraphsCount = 1;
    }
  }

  const previewLengthBytes = Buffer.byteLength(previewContent, 'utf-8');

  return {
    content: previewContent,
    isTruncated: true,
    accessGranted: false,
    paywallMeta: {
      visibility,
      teaserParagraphsCount,
      requiredTierId,
      totalLengthBytes,
      previewLengthBytes,
    },
  };
}

/**
 * Construit la description publique d'un article (SEO, OpenGraph, JSON-LD) à
 * partir du contenu DÉJÀ tronqué par `sliceContentAtPaywall`.
 *
 * ⚠️ Ne JAMAIS lui passer le contenu brut d'un article verrouillé : le passage
 * réservé se retrouverait alors dans le DOM public, dans les métadonnées
 * OpenGraph/JSON-LD et dans l'index des moteurs de recherche — soit un
 * contournement silencieux du paywall (zéro-fuite).
 */
export function buildPublicDescription(publicContent: string, maxLength = 200): string | undefined {
  const text = (publicContent || '')
    .replace(/<[^>]*>?/gm, '')
    .replace(/\s+/g, ' ')
    .trim();
  if (!text) return undefined;
  return text.slice(0, maxLength);
}

/** Entitlements d'un visiteur anonyme : aucun accès au contenu verrouillé. */
export const ANONYMOUS_ENTITLEMENTS: UserEntitlements = {
  isMember: false,
  isPaidSubscriber: false,
  tierId: null,
};

/**
 * Extrait public d'un article destiné à un visiteur ANONYME : tronque d'abord
 * le contenu au paywall (aucun entitlement), puis en dérive une description
 * texte (HTML retiré, espaces normalisés).
 *
 * ⚠️ C'est LA seule porte d'entrée autorisée pour alimenter une description
 * SEO/OpenGraph/JSON-LD ou un extrait de carte publique. Passer directement
 * `article.content` y ferait fuiter le passage réservé — dans le DOM public,
 * les métadonnées sociales et l'index des moteurs de recherche.
 *
 * Retourne `undefined` si le contenu est vide (l'appelant décide du fallback).
 */
export function buildPublicExcerpt(
  rawContent: string,
  visibility: ContentVisibility = ContentVisibility.PUBLIC,
  requiredTierId?: string | null,
  maxLength = 200
): string | undefined {
  const cut = sliceContentAtPaywall(
    rawContent || '',
    ANONYMOUS_ENTITLEMENTS,
    visibility,
    requiredTierId
  );
  return buildPublicDescription(cut.content, maxLength);
}
